FCirci Multi 1 - Model changes

hepcat72 opened the pull request:
## Summary Change Description

Almost everything changed in here relates to changes necessary to give the tracer/label appearance/disappearance rate calculations a new home so that the FCirc format output can be split up by serum sample, tracer, and element.  What was previously 1 row in the FCirc output table can now be a row for every tracer & its labels.

## Affected Issue Numbers

- Resolves #395
- Resolves #398

## Code Review Notes

The only substantial portion of these changes that are not related to the theme of this PR are the `clear update buffer` and `.using(db)` changes in the sample table loader code (which will be in a separate PR).

I turned off the CI tests for this PR.  These changes are just copies of files from the fcirc_multi branch dropped into a new branch off the `infusate_label` branch (the starting point of the master branch for this effort).  I am breaking that code review up into these smaller PRs that are separated by theme.  I am doing it by file simply because it's easier.  If any code here references something you cannot find, just consult to `fcirc_multi` branch/PR.

The PRs I will be submitting are:

- `fcirc_multi_models` (this PR)
- `fcirc_multi_format`
- `fcirc_multi_validation`
- `fcirc_multi_pagination`
- `fcirc_multi_tests`
- `fcirc_multi_minor` (comments/renames/etc)

## Checklist

These are the checks for the overall change set...

- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [ ] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [All *multi_working* tag tests pass locally](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*
  - [x] `@tag("multi_working")` added to newly passing test functions/classes *(or no newly passing tests)*
  - [x] `@tag("multi_broken")`, `@tag("multi_unknown")`, or `@tag("multi_mixed")` removed from newly passing test functions/classes *(or no newly passing tests)*

lparsons reviewed (approved):
This looks to be a solid refactor of methods. Without associated tests it's tough to determine if things work as expected, but that will get handled in the later PRs I'm guessing.

I did have a few questions and comments as I went through, but nothing blocking so I'll leave it up to Rob and approve. Nice work, and thanks for splitting things up!

hepcat72 commented on DataRepo/models/animal.py:
```diff
@@ -30,6 +29,15 @@ class Animal(HierCachedModel, ElementLabel):
     )
     infusate = models.ForeignKey(
         to="DataRepo.Infusate",
+        # PR REVIEW NOTE: This was previously RESTRICT, but I got errors during the cleanup of the validation
+        #       database that it couldn't delete some infusate records because of links to it from animal, which I
+        #       didn't expect... I thought it would delete only if there didn't exist any other links to it, and all
+        #       the animals were being deleted.  Perhaps the first time, it wouldn't delete, but the last animal that
+        #       links to it should delete the infusate...  How do I get that behavior?
+        #       Here was the error:
+        #       django.db.models.deletion.RestrictedError: ("Cannot delete some instances of model 'Infusate' because
+        #       they are referenced through restricted foreign keys: 'Animal.infusate'.", {<Animal: 090320_M1>,
+        #       <Animal: 090320_M2>, <Animal: 090320_M3>, ...
         on_delete=models.RESTRICT,
```
Right. The comment is stale.  Near the end (after I fixed the real problem), I switched it back and reran the tests and they passed. I just forgot to update/remove the comment.

I read up on RESTRICT when I changed it back and I think it's what we want. In the case of Animal:Infusate, it says essentially "Don't allow deletion of an infusate if an Animal record still links to it".  I.e. since Animal:Infusate is M:1, a cascading deletion from 1 animal record to a linked infusate could orphan another animal record from that same infusate. Restrict disallows that, but if you delete the infusate record first, then the Animal records, it will work. You just don't want to cascade *through* an animal record.

hepcat72 commented on DataRepo/models/animal.py:
```diff
@@ -94,275 +102,72 @@ class Animal(HierCachedModel, ElementLabel):
-        )
+    @property  # type: ignore
+    @cached_function
+    def tracers(self):
+        if self.infusate.tracers.count() == 0:
+            warnings.warn(f"Animal [{self.animal}] has no tracers.")
+        return self.infusate.tracers.all()
```
Huh. Yeah, I coded it to assume there was always an infusate. I thought I recalled we'd decided to make that assumption (at least for now). This would throw a NoneType error if it was none. There would probably be changes elsewhere that are necessary as well.  I think that for right now, we should change Animal.infusate to `null=False` and `blank=False` and create an issue to allow no infusate.

hepcat72 commented on DataRepo/models/animal.py:
```diff
@@ -94,275 +102,72 @@ class Animal(HierCachedModel, ElementLabel):
     @property  # type: ignore
     @cached_function
-    def final_serum_sample(self):
+    def last_serum_sample(self):
```
Yeah, I kept seeing `last` used elsewhere, especially dotted to the ORM objects, so I decided to make it consistent.

hepcat72 commented on DataRepo/models/fcirc.py:
```diff
@@ -0,0 +1,282 @@
+from DataRepo.models.hier_cached_model import HierCachedModel, cached_function
+
+
+class FCirc(HierCachedModel):
```
Yeah, I struggled with what to call this class. I decided that since it would not be user-facing and since we already call the *format* "FCirc", it didn't matter enough to figure out the best name.  Though I am open to suggestions.

- [ ] TODO: Add a docstring to the FCirc class.

jcmatese replied:
Months ago, I had assumed that the "derived_data" class hanging off peak_group would perhaps be more "generic", potentially storing multiple possible named computations, based on defined rules.  But if fcirc is the "primary"/sole number we are targetting (or even if there are only a few derivations), I am fine with this direction.

hepcat72 replied:
I don't understand what you're trying to say John. Fcirc and PeakGroupLabel are storing different things, each of which has a specific relationship with its related classes, and neither needing an argument and producing duplicate values. FCirc is under Sample and calculates unique values per serum sample, tracer, and element combo.  PeakGroupLabel calculates unique values per PeakGroup & element combo. Both are necessary where they are. There's no direction toward one or the other model.  The direction is toward making fields to store the values so that they can be searchable, and that means they need to be in these separate classes.

jcmatese replied:
Sorry, I think I was just confused about the `PeakGroup/Data` dependencies/origins of fcirc, and whether this singular calculation requires a dedicated model for it (`Fcirc`), or something more generic/extendable.  I am not suggesting any changes are required, since I have no examples of any other calculations that would be similar to fcirc(having similar relationship dependencies).

hepcat72 replied:
It was certainly confusing to figure out. Had a few false starts. But I think the current structure is correct. There might exist a better way of arranging things, but when you think about it, every table added is essentially a linking table if you think of the element fields as a link to an "Element" (or ElementLabel) table. AnimalLabel is a link between Animal and Element. PeakGroupLabel is a link between PeakGroup and Element. And FCirc is a link between Sample, Tracer, and Element. And each of these intersections between these tables has unique data associated with it.

Once I started thinking about it that way, it made more sense.

hepcat72 commented on DataRepo/models/peak_group_label.py:
```diff
@@ -188,17 +188,399 @@ def normalized_labeling(self):
+
+    @property  # type: ignore
+    @cached_function
+    def tracer(self):
+        from DataRepo.models.tracer import Tracer
+
+        try:
+            # This gets the tracer for this peakgroup (based on the compounds)
+            this_tracer = self.animal.infusate.tracers.get(
+                compound__id__in=list(
+                    self.peak_group.compounds.values_list("id", flat=True)
+                ),
+            )
+        except Tracer.DoesNotExist:
+            this_tracer = None
+
+        return this_tracer
```
Yeah, that's right.

- [ ] TODO: Add a doc string to `PeakGroupLabel.tracer`

hepcat72 replied:
While I have no doubt that there may exist the opportunity for some consolidation, I'm not quite sure what consolidation you're suggesting. Could you name the methods you think could be consolidated?

hepcat72 replied:
By "here", I assume you mean inside `PeakGroupLabel`. If that's the case, then I can say that I did consider that.  In fact, step back enough commits in the PR where I introduced AnimalLabel, and you can see the point where I had tried it out. AnimalLabel exists to provide a home for methods whose relationship with any individual animal is 1 record per labeled element among all the tracers in an infusate. There do exist queries where you can put them inside PeakGroupLabel, however with caching, it would be re-computing and repeatedly caching the same value for every related record.  That would be a lot of redundancy.  On the other hand, in AnimalLabel, every record is a unique result for those methods.

Another reason they exist where they do is so that you do not have to provide an argument.  I.e., I could have put them in `Animal`, but requiring an argument makes the caching more complicated and I didn't feel that was worth it.  The only real method that dictates the existence of this class is `serum_tracers_enrichment_fraction`.  The other methods (`tracers` and `last_serum_tracer_label_peak_groups`) are just there to assist `serum_tracers_enrichment_fraction` without having to supply arguments (so they can be easily cached).

hepcat72 commented on DataRepo/utils/sample_table_loader.py:
```diff
@@ -149,7 +151,7 @@ def load_sample_table(self, data, skip_researcher_check=False, debug=False):
         animals_to_uncache = []
 
-        # Create a list to hold the csv reader data so that iterations from validating doesn't leave the csv reader
+        # Create a list to hold the csv reader data so that iterations from validating cleardoesn't leave the csv reader
```
Yeah.  Probably didn't knwo what window I was typing in.

- [ ] TODO: clear typo

fkang-pu reviewed (commented):
Just submit my comments while spending more time to understand a list of methods.

fkang-pu reviewed (approved):
Rob has done a huge amount of work with model changes to handle cases with multiple tracers/elements. Very impressive! I reviewed the methods for each model and spot checked a few peakgroups with querysets and viewed some pages in my local branch. But it's bit hard for me to picture details for Fcirc values without reviewing other PRs.  I agree with Lance that we should rollout Rob's PRs and my PR into dev. server ASAP, which would make easier for us to view the pages and evaluate cases in details. Then we can implement improvements in future PRs.

Here are my suggestions for future work:

1. select a short list of animals and peakgroups to build a testset. Currently the tests run different loadings and took a long time. 
e.g.
peakgroups covering:
* tracer peakgroups (e.g. tracer_group_name= "BCAAs (VLI)" and measured compound="valine")
* peakgroups linked to multiple compounds (e.g. citrate/isocitrate)
* animals with different serum sample cases (no serum sample, one serum sample, multiple serum samples)
* at least one animal case with expected Fcirc values for multiple tracers/elements.

2. store additional values in PeakGroupLabel during data loading.
e.g. Total Abundance; Enrichment Fraction; Enrichment Abundance

3. investigate if a custom manager for PeakGroupLabel can simplify some methods (e.g. get tracer_peakgroups per animal, or per sample, or per serum samples)

possible solution with queryset(s):
```
from django.db.models import F
from DataRepo.models import * 
columns = [
    "id",
    "peak_group_id",
    "element", 
    "peak_group__compounds__id",
    "peak_group__compounds__name",
    "peak_group__formula",
    "peak_group__name",
    "peak_group__msrun",
    "peak_group__msrun__sample__id",
    "peak_group__msrun__sample__name",
    "peak_group__msrun__sample__tissue__name",
    "peak_group__msrun__sample__animal__id",
    "peak_group__msrun__sample__animal__name",
    "peak_group__msrun__sample__animal__infusion_rate",
    "peak_group__msrun__sample__animal__body_weight",
    "peak_group__msrun__sample__animal__infusate__name",
    "peak_group__msrun__sample__time_collected"
]
* get all peakgroups with selected fields
pg_all = PeakGroupLabel.objects.select_related().all().values(*columns)

* get peakgroups per animal (e.g. xz1078)
pg_per_animal=pg_all.filter(peak_group__msrun__sample__animal__name="xz1078")

* get tracer peakgroups per animal (no need to loop through tracers)
tracer_pg_per_animal = pg_per_animal.filter(peak_group__msrun__sample__animal__infusate__name__contains=F("peak_group__compounds__name"))

* tracer peakgrousp for serum samples per animal
tracer_pg_per_animal_serum_samples = tracer_pg_per_animal.filter(peak_group__msrun__sample__tissue__name__icontains="serum")
```

hepcat72 commented:
> Rob has done a huge amount of work with model changes to handle cases with multiple tracers/elements. Very impressive! I reviewed the methods for each model and spot checked a few peakgroups with querysets and viewed some pages in my local branch. But it's bit hard for me to picture details for Fcirc values without reviewing other PRs. I agree with Lance that we should rollout Rob's PRs and my PR into dev. server ASAP, which would make easier for us to view the pages and evaluate cases in details. Then we can implement improvements in future PRs.
> 
> Here are my suggestions for future work:
> 
> 1. select a short list of animals and peakgroups to build a testset. Currently the tests run different loadings and took a long time.
>    e.g.
>    peakgroups covering:
> 
> * tracer peakgroups (e.g. tracer_group_name= "BCAAs (VLI)" and measured compound="valine")
> * peakgroups linked to multiple compounds (e.g. citrate/isocitrate)
> * animals with different serum sample cases (no serum sample, one serum sample, multiple serum samples)
> * at least one animal case with expected Fcirc values for multiple tracers/elements.

There is an issue for this: #480.

I invite you to pick that issue up and create a PR.

> 2. store additional values in PeakGroupLabel during data loading.
>    e.g. Total Abundance; Enrichment Fraction; Enrichment Abundance

I disagree.  While `enrichment_fraction` and `enrichment_abundance` are already in `PeakGroupLabel`, `total_abundance` should live in `PeakGroup`.  If it lived in `PeakGroupLabel`, it would have to traverse 2 related models to compute the value and that value would be duplicated for every label.  In `PeakGroup`, the value is unique and exists at the common intersection of all the peak data records.  I don't see what the benefit of moving it into `PeakGroupLabel` would be.

> 3. investigate if a custom manager for PeakGroupLabel can simplify some methods (e.g. get tracer_peakgroups per animal, or per sample, or per serum samples)
> 
> possible solution with queryset(s):
> 
> ```
> from django.db.models import F
> from DataRepo.models import * 
> columns = [
>     "id",
>     "peak_group_id",
>     "element", 
>     "peak_group__compounds__id",
>     "peak_group__compounds__name",
>     "peak_group__formula",
>     "peak_group__name",
>     "peak_group__msrun",
>     "peak_group__msrun__sample__id",
>     "peak_group__msrun__sample__name",
>     "peak_group__msrun__sample__tissue__name",
>     "peak_group__msrun__sample__animal__id",
>     "peak_group__msrun__sample__animal__name",
>     "peak_group__msrun__sample__animal__infusion_rate",
>     "peak_group__msrun__sample__animal__body_weight",
>     "peak_group__msrun__sample__animal__infusate__name",
>     "peak_group__msrun__sample__time_collected"
> ]
> * get all peakgroups with selected fields
> pg_all = PeakGroupLabel.objects.select_related().all().values(*columns)
> 
> * get peakgroups per animal (e.g. xz1078)
> pg_per_animal=pg_all.filter(peak_group__msrun__sample__animal__name="xz1078")
> 
> * get tracer peakgroups per animal (no need to loop through tracers)
> tracer_pg_per_animal = pg_per_animal.filter(peak_group__msrun__sample__animal__infusate__name__contains=F("peak_group__compounds__name"))
> 
> * tracer peakgrousp for serum samples per animal
> tracer_pg_per_animal_serum_samples = tracer_pg_per_animal.filter(peak_group__msrun__sample__tissue__name__icontains="serum")
> ```

I cannot find the `tracer_peakgroups` method you are referencing, but I assume you're referring to the following methods, and I will explain why they were defined where they were and why that's a good idea.

- `Animal.last_serum_tracer_peak_groups` - This obtains the last peak group for every tracer, whether it's in the last serum sample or a previous serum sample.
- `AnimalLabel.last_serum_tracer_label_peak_groups` - This uses `Animal.last_serum_tracer_peak_groups`, but filters it for tracers containing `AnimalLabel.element`.
- `Sample.last_tracer_peak_groups` - This obtains the last peak group for every tracer for a specific serum sample (whether it's the last one or not).
- `Sample.peak_groups(compound)` - Currently unused and can be deleted. Takes a compound argument. Can't think of any reason to keep it.
- `FCirc.last_peak_group` - This uses `Sample.last_tracer_peak_groups` (which is cached) to obtain the last peak groups and filters them for `FCirc.tracer`.  In my latest branch, I have renamed this and added another method, which now are:
  - `FCirc.last_peak_group_in_sample` which calls `self.serum_sample.last_tracer_peak_groups` and filters it for `FCirc.tracer`
  - `FCirc.last_peak_group_in_animal` which calls `self.serum_sample.animal.last_serum_tracer_peak_groups` and filters it for `FCirc.tracer`
  - These methods are used in another method called `FCirc.is_last_serum_peak_group` which basically returns `self.last_peak_group_in_sample == self.last_peak_group_in_animal` to indicate whether the peak group that record represents is the last such peak group or not.  One could argue in this case that the `_in_sample` and `_in_animal` methods could be deleted in do them both inside the `FCirc.is_last_serum_peak_group` method, but I think having them makes the code clearer.

Each method has a distinct purpose/result and lives in a class where it does not require an argument and where its return value is unique.  I see no benefit to moving them that is not outweighed by the benefit of having them where they are.

Speaking of the benefits, note that the [plan](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/460) for many of the currently cached methods is to create maintained fields so that they are **searchable**.  And whether those methods are cached or used to update maintained fields, they should exist where they do not require arguments and where the values they output are unique.

jcmatese commented on DataRepo/models/animal_label.py:
```diff
@@ -36,149 +41,146 @@ class Meta:
+            # For every tracer whose compound contains this element
+            for tracer in self.tracers.all():
+
+                # PR REVIEW NOTE: This is essentially how this was calculated before... but I wonder if the
+                # total_atom_count should be the number of labeled atoms originally in the tracer.  I.e. if it wasn't
+                # fully labeled, the enrichment would never (likely) be 100%.  Is that what we *want*?
```
I think this is a good point, because I don't _remember_ partial labelling was considered previously?  Might want to make this a followup issue, to consult with researchers, and revisit at a later time (add another method, or clarify naming)?

hepcat72 commented:
Closing as approved, but nothing to merge here.  Merging will happen in #493
